transport/client.go

```diff
@@ -0,0 +1,316 @@
+		return ca.NewCFSSLProvider(id, nil)
+	}
+)
+
+// A Transport is capable of providing transport-layer security using
+// TLS.
+type Transport struct {
+	// Before defines how long before the certificate expires the
+	// transport should start attempting to refresh the
+	// certificate. For example, if this is 24h, then 24 hours
+	// before the certificate expires the Transport will start
+	// attempting to replace it.
+	Before time.Duration
+
+	// Provider contains a key management provider.
+	Provider kp.KeyProvider
+
+	// CA contains a mechanism for obtaining signed certificates.
+	CA ca.CertificateAuthority
+
+	// TrustStore contains the certificates trusted by this
+	// transport.
+	TrustStore *roots.TrustStore
+
+	// ClientTrustStore contains the certificate authorities to
+	// use in verifying client authentication certificates.
+	ClientTrustStore *roots.TrustStore
+
+	// Identity contains information about the entity that will be
+	// used to construct certificates.
+	Identity *core.Identity
+}
+
+// TLSClientAuthClientConfig returns a new client authentication TLS
+// configuration that can be used for a client using client auth
+// connecting to the named host.
+func (tr *Transport) TLSClientAuthClientConfig(host string) (*tls.Config, error) {
+	cert, err := tr.getCertificate()
+	if err != nil {
+		return nil, err
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      tr.TrustStore.Pool(),
+		ServerName:   host,
+		CipherSuites: core.CipherSuites,
+		MinVersion:   tls.VersionTLS12,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+	}, nil
+}
+
+// TLSClientAuthServerConfig returns a new client authentication TLS
+// configuration for servers expecting mutually authenticated
+// clients. The clientAuth parameter should contain the root pool used
+// to authenticate clients.
+func (tr *Transport) TLSClientAuthServerConfig() (*tls.Config, error) {
+	cert, err := tr.getCertificate()
+	if err != nil {
+		return nil, err
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      tr.TrustStore.Pool(),
+		ClientCAs:    tr.ClientTrustStore.Pool(),
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		CipherSuites: core.CipherSuites,
+		MinVersion:   tls.VersionTLS12,
+	}, nil
+}
+
+// TLSServerConfig is a general server configuration that should be
+// used for non-client authentication purposes, such as HTTPS.
+func (tr *Transport) TLSServerConfig() (*tls.Config, error) {
+	cert, err := tr.getCertificate()
+	if err != nil {
+		return nil, err
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		CipherSuites: core.CipherSuites,
+		MinVersion:   tls.VersionTLS12,
+	}, nil
+}
+
+// New builds a new transport from an identity and a before time. The
+// before time tells the transport how long before the certificate
+// expires to start attempting to update when auto-updating.
+func New(before time.Duration, identity *core.Identity) (*Transport, error) {
```

Comment:
If the auto updater is running, every time it checks the time (e.g. every `PollInterval` seconds) it will request a new certificate.